### PR TITLE
[USM] Add HTTP2 decoding for TLS traffic

### PR DESCRIPTION
**Pull request by Yumasi**
### What does this PR do?

This PR makes USM able to decode the HTTP2 traffic from our TLS implementation.

**This is a draft PR**

- [ ] Unit tests

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/ebpf/c/protocols/http2/decoding-defs.h [hunk 1]
```diff
@@ -8,7 +8,8 @@
 // Maximum number of frames to be processed in a single TCP packet. That's also the number of tail calls we'll have.
 // NOTE: we may need to revisit this const if we need to capture more connections.
 #define HTTP2_MAX_FRAMES_ITERATIONS 10
-#define HTTP2_MAX_FRAMES_TO_FILTER  500
+/* #define HTTP2_MAX_FRAMES_TO_FILTER  450 */
+#define HTTP2_MAX_FRAMES_TO_FILTER 200
```
Comment (Yumasi) [inline pkg/network/ebpf/c/protocols/http2/decoding-defs.h:12]: Note to reviewers: this was changed because of instruction count issues, and the value is way lower than it needs to be so I didn't have to change it every time I added a log line. Once the PR gets close to approval, I shall figure out what the new maximum value is (and maybe introduce a separate one for TLS).

File: pkg/network/ebpf/c/protocols/classification/defs.h [hunk 2]
```diff
@@ -135,6 +135,7 @@ typedef enum {
     PROG_HTTP,
     PROG_HTTP2,
     PROG_HTTP2_FRAME_PARSER,
+    PROG_HTTP2_TLS_ENTRY,
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/classification/defs.h:138]: Not sure why we need that program in the socket filter
can you elaborate?

File: pkg/network/ebpf/c/protocols/classification/structs.h [hunk 3]
```diff
@@ -27,6 +27,8 @@ typedef struct {
     conn_tuple_t tup;
     __u64 tags;
     char *buffer_ptr;
+    size_t len;
+    size_t off;
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/classification/structs.h:31]: Let's have inline comments for documentation

File: pkg/network/ebpf/c/protocols/http/buffer.h [hunk 4]
```diff
@@ -17,19 +17,35 @@
...
         if (ret >= 0) {                                                                                                 \
             return;                                                                                                     \
         }                                                                                                               \
         const __u64 read_size_until_end_of_page = PAGESIZE - ((__u64)src % PAGESIZE);                                   \
         const __u64 size_to_read = read_size_until_end_of_page < total_size ? read_size_until_end_of_page : total_size; \
-        bpf_probe_read_user_with_telemetry(dst, size_to_read, src);                                                     \
+        bpf_probe_read_user(dst, size_to_read, src);                                                                    \
         return;                                                                                                         \
-    }                                                                                                                   \
+    }
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http/buffer.h:28]: That change should be documented in the PR description
ps, We can separate the http2 implementation from http, by adding the "bpf probe read" function as another argument

File: pkg/network/ebpf/c/protocols/http/buffer.h [hunk 5]
```diff
@@ -17,19 +17,35 @@
...
+// note: arm64 bpf_probe_read_user() could page fault if the CLASSIFICATION_MAX_BUFFER overlap a page
+#pragma unroll(CLASSIFICATION_MAX_BUFFER - 1)
+        for (int i = 0; i < CLASSIFICATION_MAX_BUFFER - 1; i++) {
+            bpf_probe_read_user(&buffer[i], 1, &data[i]);
+            if (buffer[i] == 0) {
+                return;
+            }
+        }
+    }
+}
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http/buffer.h:47]: Why is it needed? we have it defined in line 31 (read_into_user_buffer_classification)

File: pkg/network/ebpf/c/protocols/http/http.h
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http/http.h:1]: I don't see actual changes here, let's revert it to keep the PR small and focused as possible

File: pkg/network/ebpf/c/protocols/http2/decoding-common.h [hunk 6]
```diff
@@ -0,0 +1,109 @@
+#ifndef DECODING_COMMON_H_
+#define DECODING_COMMON_H_
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/decoding-common.h:2]:
```suggestion
#ifndef __DECODING_COMMON_H_
#define __DECODING_COMMON_H_
```

File: pkg/network/tracer/testutil/grpc/testdata/cert.pem.0
Comment (guyarb) [inline pkg/network/tracer/testutil/grpc/testdata/cert.pem.0:1]: We already have test certificate and a key (`pkg/network/protocols/http/testutil/testdata`)
I do recommend to use them instead of having a duplication

File: pkg/network/tracer/testutil/grpc/server.go [hunk 7]
```diff
@@ -13,12 +13,16 @@ import (
 	"log"
 	"math"
 	"net"
+	"path/filepath"
 	"sync"
 	"time"
 
+	httputil "github.com/DataDog/datadog-agent/pkg/network/protocols/http/testutil"
```
Comment (guyarb) [inline pkg/network/tracer/testutil/grpc/server.go:20]: local imports should be the last group

File: pkg/network/tracer/testutil/grpc/server.go [hunk 8]
```diff
@@ -220,15 +224,33 @@ func serialize(point *routeguide.Point) string {
 }
 
 // NewServer returns a new instance of the gRPC server.
-func NewServer(addr string) (*Server, error) {
+func NewServer(addr string, enableTls bool) (*Server, error) {
```
Comment (guyarb) [inline pkg/network/tracer/testutil/grpc/server.go:227]: rename `Tls` to `TLS`

File: pkg/network/tracer/testutil/grpc/client.go [hunk 9]
```diff
@@ -140,13 +143,20 @@ type Options struct {
 	DialTimeout time.Duration
 	// CustomDialer allows to modify the underlying dialer used by grpc package. Set nil for the default dialer.
 	CustomDialer *net.Dialer
+	// WithTls makes the clients Dial using TLS
+	WithTls bool
```
Comment (guyarb) [inline pkg/network/tracer/testutil/grpc/client.go:147]: rename to `WithTLS`

File: pkg/network/tracer/testutil/grpc/client.go [hunk 10]
```diff
@@ -140,13 +143,20 @@ type Options struct {
...
 func NewClient(addr string, options Options) (Client, error) {
-	gRPCOptions := []grpc.DialOption{
-		grpc.WithTransportCredentials(insecure.NewCredentials()),
+	gRPCOptions := []grpc.DialOption{}
+
+	if options.WithTls {
+		gRPCOptions = append(gRPCOptions, grpc.WithTransportCredentials(credentials.NewTLS(&tls.Config{ServerName: "", InsecureSkipVerify: true})))
+	} else {
+		gRPCOptions = append(gRPCOptions, grpc.WithTransportCredentials(insecure.NewCredentials()))
 	}
```
Comment (guyarb) [inline pkg/network/tracer/testutil/grpc/client.go:158]:
```suggestion
        creds := grpc.WithTransportCredentials(insecure.NewCredentials())
	if options.WithTls {
		creds = credentials.NewTLS(&tls.Config{ServerName: "", InsecureSkipVerify: true})
	}
	gRPCOptions = append(gRPCOptions, grpc.WithTransportCredentials(creds))
```

File: pkg/network/protocols/http2/types_linux.go [hunk 11]
```diff
@@ -36,6 +36,7 @@ type EbpfTx struct {
 	Tup                   connTuple
 	Response_last_seen    uint64
 	Request_started       uint64
+	Tags                  uint64
```
Comment (guyarb) [inline pkg/network/protocols/http2/types_linux.go:39]: Please measure the impact of that change to USM program and document it in the pr description 
changing the http2_transaction struct impacts the size of the http2 kernel maps (hence system-probe memory)

File: pkg/network/protocols/http2/protocol.go [hunk 12]
```diff
@@ -202,6 +227,7 @@ func (p *protocol) DumpMaps(output *strings.Builder, mapName string, currentMap
 
 func (p *protocol) processHTTP2(data []byte) {
 	tx := (*EbpfTx)(unsafe.Pointer(&data[0]))
+	log.Debugf("Statig tags: %v", tx.StaticTags())
```
Comment (guyarb) [inline pkg/network/protocols/http2/protocol.go:230]: should it be in a debug log?

File: pkg/network/ebpf/c/runtime/usm.c
Comment (guyarb) [inline pkg/network/ebpf/c/runtime/usm.c:1]: All changes are not related to that branch, please revert

File: pkg/network/ebpf/c/protocols/tls/native-tls.h
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/tls/native-tls.h:1]: All changes are not related to that branch, please revert

File: pkg/network/ebpf/c/protocols/tls/java/erpc_handlers.h
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/tls/java/erpc_handlers.h:1]: All changes are not related to that branch, please revert

File: pkg/network/ebpf/c/protocols/tls/https-maps.h [hunk 13]
```diff
@@ -0,0 +1,11 @@
+#ifndef HTTPS_MAPS_H_
+#define HTTPS_MAPS_H_
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/tls/https-maps.h:2]:
```suggestion
#ifndef __HTTPS_MAPS_H_
#define __HTTPS_MAPS_H_
```

File: pkg/network/ebpf/c/protocols/tls/https-maps.h [hunk 14]
```diff
@@ -0,0 +1,11 @@
...
+
+#include "protocols/http2/decoding-defs.h"
+
+#define HTTP2_TLS_MAX_SIZE 2048
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/tls/https-maps.h:6]: unused

File: pkg/network/ebpf/c/protocols/tls/https-maps.h [hunk 15]
```diff
@@ -0,0 +1,11 @@
...
+
+BPF_HASH_MAP(http2_tls_states, http2_tls_state_key_t, http2_tls_state_t, 10)
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/tls/https-maps.h:8]: why 10?

Comment (guyarb) [inline pkg/network/ebpf/c/protocols/tls/https-maps.h:8]: please add the maps to the http2/protocol.go file

Comment (guyarb) [inline pkg/network/ebpf/c/protocols/tls/https-maps.h:8]: please document the meaning of the maps

File: pkg/network/ebpf/c/protocols/http2/decoding-defs.h [hunk 16]
```diff
@@ -121,4 +123,18 @@ typedef struct {
     http2_frame_with_offset frames_array[HTTP2_MAX_FRAMES_ITERATIONS] __attribute__((aligned(8)));
 } http2_tail_call_state_t;
 
+// TLS
+
+// This is used as a key to get the state of an http2 TLS connection.
+typedef struct {
+    conn_tuple_t tup;
+    __u32 length;
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/decoding-defs.h:131]: why is length part of the key?

File: pkg/network/ebpf/c/protocols/http2/decoding-defs.h [hunk 17]
```diff
@@ -121,4 +123,18 @@ typedef struct {
...
+} http2_tls_state_key_t;
+
+typedef struct {
+    bool relevant;
+    __u32 stream_id;
+    __u8 frame_flags;
+} http2_tls_state_t;
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/decoding-defs.h:138]: please document the struct and its field
I'm not sure the meaning of relevant

File: pkg/network/ebpf/c/protocols/http2/decoding.h [hunk 18]
```diff
@@ -82,48 +57,7 @@ static __always_inline bool read_var_int(struct __sk_buff *skb, skb_info_t *skb_
...
-    __u64 *counter_ptr = bpf_map_lookup_elem(&http2_dynamic_counter_table, tup);
-    if (counter_ptr != NULL) {
-        return counter_ptr;
-    }
-    __u64 counter = 0;
-    bpf_map_update_elem(&http2_dynamic_counter_table, tup, &counter, BPF_ANY);
-    return bpf_map_lookup_elem(&http2_dynamic_counter_table, tup);
-}
-
 // parse_field_indexed is handling the case which the header frame is part of the static table.
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/decoding.h:60]: seems out of place

File: pkg/network/ebpf/c/protocols/http2/decoding.h [hunk 19]
```diff
@@ -504,4 +407,115 @@ int socket__http2_frames_parser(struct __sk_buff *skb) {
...
+        http2_tls_state_t new_state = { 0 };
+        new_state.relevant = relevant;
+        new_state.stream_id = header.stream_id;
+        new_state.frame_flags = header.flags;
+
+        key.length = header_only ? header.length : info->len;
+
+        bpf_map_update_elem(&http2_tls_states, &key, &new_state, BPF_ANY);
+        if (header_only)
+            goto exit;
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/decoding.h:446]:
```suggestion
        if (header_only) {
            goto exit;
        }
```

File: pkg/network/ebpf/c/protocols/http2/tls-decoding.h [hunk 20]
```diff
@@ -0,0 +1,278 @@
...
+
+#include "protocols/http/buffer.h"
+#include "protocols/http2/decoding-common.h"
+#include "protocols/http2/maps-defs.h"
+#include "protocols/tls/https-maps.h"
+
+READ_INTO_USER_BUFFER(http2_preface, HTTP2_MARKER_SIZE)
+READ_INTO_USER_BUFFER(http2_frame_header, HTTP2_FRAME_HEADER_SIZE)
+READ_INTO_USER_BUFFER(http2_char, sizeof(__u8))
+READ_INTO_USER_BUFFER(path, HTTP2_MAX_PATH_LEN)
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/tls-decoding.h:15]: If you have a pattern, let's stick with it

```suggestion
READ_INTO_USER_BUFFER(http2_path, HTTP2_MAX_PATH_LEN)
```

File: pkg/network/ebpf/c/protocols/http2/tls-decoding.h [hunk 21]
```diff
@@ -0,0 +1,278 @@
+#ifndef TLS_DECODING_H_
+#define TLS_DECODING_H_
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/tls-decoding.h:2]:
```suggestion
#ifndef __TLS_DECODING_H_
#define __TLS_DECODING_H_
```

File: pkg/network/ebpf/c/protocols/http2/tls-decoding.h [hunk 22]
```diff
@@ -0,0 +1,278 @@
...
+#include "protocols/http2/decoding-common.h"
+#include "protocols/http2/maps-defs.h"
+#include "protocols/tls/https-maps.h"
+
+READ_INTO_USER_BUFFER(http2_preface, HTTP2_MARKER_SIZE)
+READ_INTO_USER_BUFFER(http2_frame_header, HTTP2_FRAME_HEADER_SIZE)
+READ_INTO_USER_BUFFER(http2_char, sizeof(__u8))
+READ_INTO_USER_BUFFER(path, HTTP2_MAX_PATH_LEN)
+
+static __always_inline void skip_preface_tls(tls_dispatcher_arguments_t *info) {
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/tls-decoding.h:17]: please add documentation

File: pkg/network/ebpf/c/protocols/http2/tls-decoding.h [hunk 23]
```diff
@@ -0,0 +1,278 @@
...
+        return false;
+    }
+    __u8 current_char_as_number = 0;
+    read_into_user_buffer_http2_char(&current_char_as_number, info->buffer_ptr + info->off);
+    info->off++;
+
+    return read_var_int_with_given_current_char_tls(info, current_char_as_number, max_number_for_bits, out);
+}
+
+static __always_inline bool is_relevant_frame_tls(tls_dispatcher_arguments_t *info, struct http2_frame *header) {
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/tls-decoding.h:68]: please document

File: pkg/network/ebpf/c/protocols/http2/tls-decoding.h [hunk 24]
```diff
@@ -0,0 +1,278 @@
...
+    }
+}
+
+static __always_inline void parse_frame_tls(tls_dispatcher_arguments_t *info, http2_ctx_t *http2_ctx, __u8 frame_flags) {
+    http2_stream_t *current_stream = http2_fetch_stream(&http2_ctx->http2_stream_key);
+    if (current_stream == NULL) {
+        return;
+    }
+
+    process_headers_frame_tls(info, current_stream, &http2_ctx->dynamic_index);
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/tls-decoding.h:269]: should be called if and only if the frame is headers
it can be data (with EOS)

Comment (guyarb) [reply pkg/network/ebpf/c/protocols/http2/tls-decoding.h:269]: if it's data, you need to modify the `data_off`

File: pkg/network/protocols/grpc/monitor_test.go [hunk 25]
```diff
@@ -45,6 +46,7 @@ func randStringRunes(n int) []rune {
 
 type USMgRPCSuite struct {
 	suite.Suite
+	withTls bool
```
Comment (guyarb) [inline pkg/network/protocols/grpc/monitor_test.go:49]:
```suggestion
	withTLS bool
```

File: pkg/network/protocols/grpc/monitor_test.go [hunk 26]
```diff
@@ -80,10 +90,17 @@ func getClientsIndex(index, totalCount int) int {
...
 	t := s.T()
+
 	cfg := config.New()
 	cfg.EnableHTTP2Monitoring = true
+	cfg.EnableGoTLSSupport = s.withTls
+	cfg.BPFDebug = true
+
+	if s.withTls {
+		log.SetupLogger(seelog.Default, "debug")
+	}
```
Comment (guyarb) [inline pkg/network/protocols/grpc/monitor_test.go:101]: Should be reverted
please see `TestMain` in `tracer_test.go` how to make `DD_LOG_LEVEL` control the log level

File: pkg/network/ebpf/c/protocols/http2/decoding.h [hunk 27]
```diff
@@ -504,4 +407,115 @@ int socket__http2_frames_parser(struct __sk_buff *skb) {
...
+        log_debug("[http2_tls_entry] could not get tls info from map");
+        return 0;
+    }
+
+    http2_tls_state_key_t key = { 0 };
+    key.tup = info->tup;
+    key.length = info->len;
+    http2_tls_state_t *state = bpf_map_lookup_elem(&http2_tls_states, &key);
+
+    if (!state) {
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/decoding.h:426]: you have a lot of indentation here
how can it changed?

having a condition like
```c
if (state && !state->relevant) {
        bpf_map_delete_elem(&http2_tls_states, &key);
        goto exit;
}

... // rest of the code
```

File: pkg/network/ebpf/c/protocols/http2/decoding.h [hunk 28]
```diff
@@ -504,4 +407,115 @@ int socket__http2_frames_parser(struct __sk_buff *skb) {
...
+        struct http2_frame header = { 0 };
+        bool relevant = is_relevant_frame_tls(info, &header);
+
+        bool header_only = info->len == HTTP2_FRAME_HEADER_SIZE;
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/decoding.h:430]: terminology here is wrong, should be `frame_header_only`

File: pkg/network/ebpf/c/protocols/http2/decoding.h [hunk 29]
```diff
@@ -504,4 +407,115 @@ int socket__http2_frames_parser(struct __sk_buff *skb) {
...
+        if (!header_only && !relevant) {
+            // The buffer seems to contain the whole frame, and the frame is not
+            // relevant: we skip it without storing some state.
+            goto exit;
+        }
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/decoding.h:435]: not sure if this condition is right
if we see get a full headers payload, then you'll drop it

while this logic fit into the scenario we've seen in go-tls grpc, does it true to grpc over python? grpc over java?

File: pkg/network/ebpf/c/protocols/http2/decoding.h [hunk 30]
```diff
@@ -504,4 +407,115 @@ int socket__http2_frames_parser(struct __sk_buff *skb) {
...
+
+        http2_tls_state_t new_state = { 0 };
+        new_state.relevant = relevant;
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/decoding.h:438]: if that's not relevant, why do you keep it?

File: pkg/network/ebpf/c/protocols/http2/decoding.h [hunk 31]
```diff
@@ -504,4 +407,115 @@ int socket__http2_frames_parser(struct __sk_buff *skb) {
...
+    http2_ctx->dynamic_index.tup = info->tup;
+    http2_ctx->http2_stream_key.stream_id = state->stream_id;
+
+    parse_frame_tls(info, http2_ctx, state->frame_flags);
+    if (info->off >= info->len) {
+        goto state_delete;
+    }
+
+state_delete:
+    bpf_map_delete_elem(&http2_tls_states, &key);
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/decoding.h:495]: you'll always go to state_delete...

**Review by guyarb — CHANGES_REQUESTED** [review]

File: pkg/network/ebpf/c/protocols/http2/decoding.h (hunk 30 above)
Comment (Yumasi) [reply pkg/network/ebpf/c/protocols/http2/decoding.h:438]: This is to avoid trying to interpret a frame content as a frame header, as doing so could add state for non existing frames or future frames that could actually be relevant.
